### PR TITLE
Merge E0002 into E0004

### LINKED ISSUE #36724: Merge E0002 into E0004
As per @arielb1's comment in https://github.com/rust-lang/rust/issues/35495 , E0002 "is basically a hack and should be merged into E0004". @jonathandturner , any comments?

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #36724.

r? @jonathandturner

### CODE REVIEW

File: src/librustc_const_eval/check_match.rs [hunk 1]
```diff
@@ -215,9 +221,10 @@ fn check_expr(cx: &mut MatchCheckCtxt, ex: &hir::Expr) {
             if inlined_arms.is_empty() {
                 if !pat_ty.is_uninhabited(cx.tcx) {
                     // We know the type is inhabited, so this must be wrong
-                    let mut err = struct_span_err!(cx.tcx.sess, ex.span, E0002,
-                                                   "non-exhaustive patterns: type {} is non-empty",
-                                                   pat_ty);
+                    let mut err = create_e0004(cx.tcx.sess, ex.span,
+                                               format!("non-exhaustive patterns: type {} \
```
Comment (nagisa) [inline src/librustc_const_eval/check_match.rs]: So now instead of having just two error codes we do allocations instead?

**Comment by jonathandturner** [comment]
Pinging @arielb1 - what did you have in mind for the refactor for E0002 and E0004?

**Comment by arielb1** [comment]
That kind of thing.

**Comment by GuillaumeGomez** [comment]
Then we're all good! :D

**Pull request merged**